File: lr/src/questions/q_0039_combination_sum.rs
pub fn combination_sum(candidates: Vec<i32>, target: i32) -> Vec<Vec<i32>> {
    let mut res = vec![];
    let mut cur = vec![];

    fn dfs(
        index: usize,
        candidates: &Vec<i32>,
        res: &mut Vec<Vec<i32>>,
        cur: &mut Vec<i32>,
        total: i32,
        target: i32,
    ) {
        if total == target {
            res.push(cur.clone());
            return;
        }
        if index >= candidates.len() || total > target {
            return;
        }

        cur.push(candidates[index]);
        dfs(
            index,
            candidates,
            res,
            cur,
            total + candidates[index],
            target,
        );

        cur.pop();
        dfs(index + 1, candidates, res, cur, total, target);
    }

    dfs(0, &candidates, &mut res, &mut cur, 0, target);

    res
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_combination_sum() {
        assert_eq!(
            combination_sum(vec![1, 2, 3], 3),
            vec![vec![1, 1, 1], vec![1, 2], vec![3]]
        )
    }
}

File: lg/src/questions/q_0039_combination_sum.go
package questions

func combinationSum(candidates []int, target int) [][]int {
	result := make([][]int, 0)
	current := make([]int, 0)

	var backtrace func(index, total int)
	backtrace = func(index, total int) {
		if total == target {
			temp := make([]int, len(current))
			copy(temp, current)
			result = append(result, temp)
			return
		}

		if index >= len(candidates) || total > target {
			return
		}

		current = append(current, candidates[index])
		backtrace(index, total+candidates[index])
		current = current[:len(current)-1]
		backtrace(index+1, total)
	}
	backtrace(0, 0)
	return result
}
